### ISSUE TITLE
Diagnostics: DataReaderDisposingEventData.ReadCount accumulates the value from different readers

### ISSUE THREAD

**Issue by arekpalinski** [dotnet/efcore#27652]
It looks that in the `DataReaderDisposingEventData.ReadCount` accumulates the values from different readers.
### Include your code

I have the following diagnostics observer:

```C#
public class MyObserver : IObserver<KeyValuePair<string, object>>
        {
            public void OnCompleted()
            {

            }

            public void OnError(Exception error)
            {

            }

            public void OnNext(KeyValuePair<string, object> item)
            {
                var value = item.Value;
                var key = item.Key;

                if (value is DataReaderDisposingEventData data)
                {
                    Console.WriteLine(data.ReadCount);
                }
            }
        }


        public class MyDiagnosticListenerObserver : IObserver<DiagnosticListener>
        {
            private readonly MyObserver _observer;

            internal MyDiagnosticListenerObserver(MyObserver observer)
            {
                this._observer = observer;
            }

            public void OnCompleted()
            {

            }

            public void OnError(Exception error)
            {

            }

            public void OnNext(DiagnosticListener value)
            {
                value.Subscribe(_observer);
            }
        }
```

Diagnostics initialization:

```C#
MyDiagnosticListenerObserver myDiagnosticListenerObserver = new MyDiagnosticListenerObserver(new MyObserver());
DiagnosticListener.AllListeners.Subscribe(myDiagnosticListenerObserver);
```

Sample code using EF (select n+1 problem):

```C#
using (var db = new Entities(conStr))
{
	foreach (var post in db.Posts.ToList())
	{
		db.Entry(post).Collection(p => p.Comments).Load();
	}
}
```

The above code writes the following (expected) output to the console when using EF Core version: 5.0.15:

```
11
5
5
5
5
5
5
5
5
5
5
```

while the same code using EF Core 6.0.3  outputs the following:

```
11
16
21
26
31
36
41
46
51
56
61
```

**Comment by arekpalinski** [comment]
I see it was tagged by `7.0.0` milestone. This is pretty important for us since this affects our product (Entity Framework Profiler  - https://hibernatingrhinos.com/products/efprof) and users get incorrect results when profiling their applications. We don't have a way to workaround this issue. Can this be considered to be fixed in `6.0.x` version?

**Comment by roji** [comment]
@arekpalinski thanks, I've verified and fixed this - as part of the EF Core 6.0 query perf optimization cycle, DbDataReader started to get recycled across executions, but the read count value wasn't properly reset.

This seems like a good candidate for servicing for 6.0 (safe one-line change), we'll discuss it.
